### DISCUSSION TITLE
wxcadm redsky and meraki audit integration

### DISCUSSION THREAD

**Discussion by drewmetz0717** [kctrey/wxcadm#6] [Q&A] [answered]
I found documents on using the meraki audit to update redsky and this function is greatly needed.  So I have been working with the code and trying to identify why it was not working for me.  I came to recognize when you import the BSSIDs it often finds duplicates in redsky and fails because meraki APs are produced with mac addresses that are very similar.  I have multiples where the Meraki manufacturing just changes the last digit of the mac. In my example E2:55:A8:0D:B8:8X is what the audit is trying to post to the BSSID list in redsky.  If you use the CSV imports available in redsky they have an field where you set masking to false and this corrects for the issue and I am able to import them via that csv import process. Has anyone else run into this? I read the documentation and I don't see anywhere in the code where I can force the audit to send the full mac of the BSSID and also send the mask False Boolean.  Anyone found a solution that I haven't?

**Comment by kctrey** [comment]
This is interesting. I don't think that wxcadm should be setting the masking for BSSIDs, but I will check the code and see. In all of my Meraki deployments, I am sending the full MAC-style BSSID without masking.

**Comment by kctrey** [comment]
I think I understand, but I want to clarify something. wxcadm doesn't attempt to set the mask value in the API call to RedSky. Are you saying that you *already* have some masked entries in RedSky, but the wxcadm audit doesn't deal well with those and ends up trying to send BSSIDs that are already known to RedSky because of the existing masked entry? That would make sense, because I never included the masking logic in the audit. I can work on doing that, but I don't have any masked entries, so I will have to create some of those and retest the audit.

If that isn't what you are saying, can you clarify? I don't see any case where wxcadm would make up a masked value to insert.

**Comment by kctrey** [comment]
I have found that "bssidMasking" gets set to True when you add a BSSID manually in the RedSky portal and is the default when adding via the API, which is why all of yours are getting set that way. RedSky doesn't document what the masking function actually does, so I have reached out to their engineers to confirm that your "X" thought is what the masking is using.  If E2:55:A8:0D:B8:8X is the actual value that the audit is seeing as the BSSID in Meraki, there is going to be a problem, since that shouldn't be a valid BSSID.

I am still working on the assumption that your RedSky already has masked entries and the audit should "expand" a masked entry to know that it includes all of the masked values. I am writing the code to handle that and will include it in the next release. Based on the answer I get from RedSky, I *may* change the logic around the bssidMasking field to not default it to True.

**Comment by kctrey** [comment]
It turns out that BSSID Masking was added by RedSky specifically to deal with Meraki's BSSID formatting. I have changed all of the relevant code in the MerakiNetwork.redsky_audit() to ensure that all BSSIDs that get added have the bssidmasking=False set on the RedSky side, which should resolve your issue. The change was delivered in v4.2.0, which is live now.

**Comment by drewmetz0717** [comment]
Trey, this is a lot closer to working now. I had some time to test it out this afternoon.  I am hitting a new issue.  If the switch is already added in redsky then the code stops running with an error. I tested this on BSSID as well and it also fails. As a workaround I pulled out the tag on the switch in the meraki dashboard and the code works. I did the same test on an Existing AP and I get the same error that the BSSID exists and the code stops. 

Is there an answer to this in the code that I can allow the existing devices to be skipped and to allow the code to continue to execute on the remaining audit?

Thanks,

Drew

**Comment by kctrey** [comment]
Can you send me, either here or privately, the exception/error that you see when it hits the existing switch? I thought that I had taken care of most of that a while ago, but I must have missed something. It is supposed to detect only the things that are missing, but if they don't match what it is expecting, it is probably trying to add them anyway.

**Reply by kctrey** [reply]
If you enable logging in your script, or just do wxcadm.console_logging() before you run the audit, the log might help as well. The code checks to see if the Location can be determined from the Meraki entry, then it makes sure the Location exists in RedSky. Only then does it look to see if the switch chassis is already in RedSky. If it isn't, it adds it, and if it is, it updates it to the correct Location. It would be helpful to know if the code is actually finding the existing Chassis or not.

**Comment by drewmetz0717** [comment]
You are right. It doesn't discover the device mac when it runs the audit so it attempts to add it to redsky.  I removed the chassis from redsky for good measure and then I ran the code twice just to be sure I didn't have any errors.  On the first run it puts the new device into the location.  The second run of the code errors out when it attempts to add the device to the location but it's already present. 

directly sent logs to your gmail account.

**Comment by kctrey** [comment]
Assuming you are using something like my sample script, can you add a print line after you attach the RedSky to the Meraki Network and send it to me privately, or you can look at it yourself and see if the MAC of the Chassis truly is in the LLDP Discovery that RedSky has? Here is an example script you can use:

``` python
import wxcadm

meraki_dashboard_token = "Your Meraki API Token"
meraki_org_name = 'Your Meraki Org Name'
meraki_network_name = 'Your Meraki Network Name'
redsky_username = 'Your Redsky Username'
redsky_password = 'Your RedSky Password'

meraki = wxcadm.Meraki(meraki_dashboard_token)
org: wxcadm.meraki.MerakiOrg = meraki.get_org_by_name(meraki_org_name)
network: wxcadm.meraki.MerakiNetwork = org.get_network_by_name(meraki_network_name)

network.attach_redsky(redsky_username, redsky_password)
print(network._redsky.get_lldp_discovery())
wxcadm.console_logging("debug")
results = network.redsky_audit(simulate=True)
print(vars(results))
```

**Comment by kctrey** [comment]
Thinking about this, it may be a pagination issue with the RedSky API. I have never actually tested this on a RedSky organization that has hundreds or thousands of entries. I wonder if I am not collecting all of the entries when I get the data from RedSky. If that's the case, I assume you *won't* see the Chassis MAC in the print from that script. I am going back through my code to make sure I am following the pagination they use on their APIs (and that they are consistent with it).

**Reply by drewmetz0717** [reply]
I sent you the logs to your gmail.  That does seem a possibility.

**Comment by kctrey** [comment, marked as answer]
That is definitely what was causing it. I have just deployed v4.2.1 which has the fix for this, and other places where I found that I wasn't handling paginated responses.

**Reply by drewmetz0717** [reply]
This new version is working on my new site which is awesome to report.  Thanks for you hard work on this Trey!! Later this month I will test this out on one of my larger sites and I will report back once I do.
